Signup: Preload the next step chunk in the flow

spen opened the pull request:
#### Changes proposed in this Pull Request

This PR aims to load the step component that would be next in the current flow ahead of time by calling the async import as soon as the current step changes.

To do so, I've done the following:

- Added a `getNextStepNameInFlow` function, allowing us to give a step name and a flow name, and get the name of the next step in that flow.
- Added and used a `preloadNextStep` function that grabs the next step name and async loads the module in both the `componentDidMount` and `componentDidUpdate` methods.
- Updated the now incorrect docs

#### Testing instructions

- Filter network tab by '`async-load-signup`'
- While signed out, start at: http://calypso.localhost:3000/start/onboarding
	- See that the first step component still loads as expected.
	- You should notice two `async-load-signup` javascript hits in the network tab - one for the first task (`async-load-signup-steps-user`) and one for the next task (`async-load-signup-steps-site-topic`).
- Continue through the steps and watch for those `async-load-signup` javascript hits - each step you progress to should result in the next step being loaded.

- Apply the same test logic to other flows - we should see the `async-load-signup...` chunks load one step before they're actually needed. (http://calypso.localhost:3000/start/import/from-url for example)

jsnajdr commented on client/signup/config/step-components.js:
```diff
@@ -2,70 +2,71 @@
+		/* webpackChunkName: "async-load-signup-steps-[request]", webpackInclude: /signup\/steps\/[a-z-]+\/index.jsx$/ */ `signup/steps/${ stepName }`
+	);
+
+const loadStepComponent = memoize( stepName => React.lazy( () => asyncLoadStep( stepName ) ) );
```
There's a bug here: the module loaded shouldn't be `stepName`, but `stepNameToFilenameMap[ stepName ]`. After fixing, the `getStepComponent` and `loadStepComponent` functions should be the same.

fditrapani commented:
I took this for a quick run and although it's working better I'm still getting the flash at times. This wouldn't be a blocker for the PR and can probably be layered in another PR but would we be able to hold on the current step, with a progress indicator (I'm thinking of the [button busy](https://wpcalypso.wordpress.com/devdocs/design/button) variety), until the new step has loaded?

fditrapani commented:
Also cc'ing @scruffian for any thoughts/input?

ramonjd commented on client/signup/main.jsx:
```diff
@@ -298,6 +301,11 @@ class Signup extends React.Component {
 
+	preloadNextStep( stepName ) {
+		const nextStep = flows.getNextStepNameInFlow( this.props.flowName, stepName );
+		nextStep && asyncLoadStep( nextStep );
```
`asyncLoadStep` expects a file name, not a step name, so I tested this PR by editing the following:

```
import { getStepComponent, asyncLoadStep, stepNameToFilenameMap } from './config/step-components';
....

nextStep && asyncLoadStep( get( stepNameToFilenameMap, nextStep ) );
```

Works well! The steps load and, at least for me, works pretty well. ❤️ I throttled to 3G and it was a marked improvement.

I'm not sure how to square the above with @jsnajdr 's [comments](https://github.com/Automattic/wp-calypso/pull/32718/files#r280109607), but once we work out how to pass the file name vs the step name between these three methods, then I think it'll work nicely.

ramonjd commented on client/signup/config/step-components.js:
```diff
@@ -2,70 +2,71 @@
+	'site-title-with-preview': 'site-title',
 };
+
+export const asyncLoadStep = stepName =>
```
Since this method builds a file path, could we rename `stepName` to `fileName` or something similar?

ramonjd commented on client/signup/config/flows.js:
```diff
@@ -152,6 +152,16 @@ const Flows = {
+		const nextStepName = get( flowSteps, nextIndex, last( flowSteps ) );
+
+		return nextStepName;
+	},
```
Any reason to import and use the extra lodash methods here? 

It's not a huge deal, and I totally understand if you're relying on lodash's fallbacks, but the app will fail pretty hard anyway if the flow objects are missing a steps array.

This works,

```
	getNextStepNameInFlow( flowName, currentStepName = '' ) {
		const flowSteps = Flows.getFlows()[ flowName ].steps;
		const nextIndex = flowSteps.indexOf( currentStepName ) + 1;
		return flowSteps[ nextIndex || flowSteps[ flowSteps.length - 1 ] ];
	},
```

Or, if you want to be defensive:

```
getNextStepNameInFlow( flowName, currentStepName = '' ) {
		const flow = Flows.getFlows()[ flowName ] || {};
		if ( flow && Array.isArray( flow.steps ) ) {
			const nextIndex = flow.steps.indexOf( currentStepName ) + 1;
			return flow.steps[ nextIndex || flow.steps[ flow.steps.length - 1 ] ];
		}
		return;
}
```

Please feel free to ignore this comment... :)

spen replied:
My thinking is that we could eventually utilise `_.get`s third argument and provide some default here. You're right though, if we're failing to access properties with dot notation, we're already failing elsewhere!

ramonjd replied:
In the interests of disclosure, I use `_.get` all the time, and think it would make a nice t-shirt logo.

spen replied:
`_.get( 'real' )` 😄

spen commented:
Hey folks thanks for taking a look at this!

I've rebased and reshaped this PR a little to fit well with @jsnajdr's changes from #32752.
Those changes allowed this PR to become a lot slimmer and more focused. It now handles pretty much just the preloading of the next step component :)

I've updated the description to better reflect these changes too.

jblz commented on client/signup/main.jsx:
```diff
@@ -300,6 +306,13 @@ class Signup extends React.Component {
+		const currentStepName = this.props.stepName;
+		const nextStepName = flows.getNextStepNameInFlow( this.props.flowName, currentStepName );
+
+		nextStepName && getStepComponent( nextStepName );
```
I wonder if it's worth `defer`ing the `getStepComponent` call here 🤔

spen replied:
Good call! That way we're not blocking anything right off the bat 👌

jsnajdr replied:
I wouldn't defer unless there is a serious reason to do it. `getStepComponent` just initiates a dynamic import. In webpack, that means checking if the module is already loaded or loading (lookup in the module map) and inserting a `<script>` tag with `onload` and `onerror` handlers if it isn't. That should be all pretty fast. The heavy lifting (parsing and executing the module code) is already async.

To answer an earlier question (from approx week ago): I agree that preloading the next step is a great followup to the previous work on async-loading steps. It should eliminate almost all instances where the user experiences an awkward waiting period when moving to the next step.

spen replied:
> getStepComponent just initiates a dynamic import.

Oh yeah, that makes sense! and thanks for those details, I'd never gone that far through with how exactly webpack handles asyncs so this is good to know :)

spen commented:
> Forking into a flow that doesn't share the same immediate step as the others.

This did cross my mind too. I wonder what the best approach would be here - should we aim to load each potential fork? That could be achievable but definitely another extra layer of complexity.
Or should we allow those few cases to not be preloaded, but better handle the loading state so that it still feels quick?
Maybe both?

Thanks for reviewing folks, and for your input - much appreciated! :)
